src/utils/stackedUtils.ts

```diff
@@ -0,0 +1,179 @@
+///<reference path="../reference.ts" />
+
+module Plottable {
+  export module Utils {
+    type StackedDatum = {
+      key: any;
+      value: number;
+      offset?: number;
+    };
+
+    export class Stacked {
+
+      /**
+       * Calculates the offset of each piece of data, in each dataset, relative to the baseline,
+       * for drawing purposes.
+       *
+       * @return {Utils.Map<Dataset, D3.Map<number>>} A map from each dataset to the offset of each datapoint
+       */
+      public static computeStackOffsets(datasets: Dataset[], keyAccessor: Accessor<any>, valueAccessor: Accessor<any>) {
+        var domainKeys = this.getDomainKeys(datasets, keyAccessor);
+
+        var dataMapArray = this._generateDefaultMapArray(datasets, keyAccessor, valueAccessor, domainKeys);
+
+        var positiveDataMapArray: D3.Map<StackedDatum>[] = dataMapArray.map((dataMap) => {
+          return Utils.Methods.populateMap(domainKeys, (domainKey) => {
+            return { key: domainKey, value: Math.max(0, dataMap.get(domainKey).value) || 0 };
+          });
+        });
+
+        var negativeDataMapArray: D3.Map<StackedDatum>[] = dataMapArray.map((dataMap) => {
+          return Utils.Methods.populateMap(domainKeys, (domainKey) => {
+            return { key: domainKey, value: Math.min(dataMap.get(domainKey).value, 0) || 0 };
+          });
+        });
+
+        var stackOffsets = this._generateStackOffsets(
+          datasets,
+          this._stack(positiveDataMapArray, domainKeys),
+          this._stack(negativeDataMapArray, domainKeys),
+          keyAccessor,
+          valueAccessor);
+
+        return stackOffsets;
+      }
+
+      /**
+       * Calculates an extent across all datasets. The extent is a <number> interval that
+       * accounts for the fact that stacked bits have to be added together when calculating the extent
+       *
+       * @return {[number]} The extent that spans all the stacked data
+       */
+      public static computeStackExtent(
+          datasets: Dataset[],
+          keyAccessor: Accessor<any>,
+          valueAccessor: Accessor<any>,
+          stackOffsets: Utils.Map<Dataset, D3.Map<number>>,
+          filter: Accessor<boolean>) {
+
+        var maxStackExtent = Utils.Methods.max<Dataset, number>(datasets, (dataset: Dataset) => {
+          var data = dataset.data();
+          if (filter != null) {
+            data = data.filter((d, i) => filter(d, i, dataset));
+          }
+          return Utils.Methods.max<any, number>(data, (datum: any, i: number) => {
+            return +valueAccessor(datum, i, dataset) +
+              stackOffsets.get(dataset).get(String(keyAccessor(datum, i, dataset)));
```

Comment:
The `String` coercion here is probably a high reason to use `Utils.Map` instead of `d3.map`